widget/infprogressbar.go

```diff
@@ -0,0 +1,170 @@
+		// slightly increase width
+		newBoxSize := fyne.Size{Width: barWidth + stepSize, Height: progressSize.Height}
+		p.bar.Resize(newBoxSize)
+	} else {
+		// move bar to the right by stepSize
+		barPos.X += stepSize
+
+		if barWidth <= minBarWidth {
+			// loop around to start when bar goes to end
+			barPos.X = 0
+			stepSize = 0
+			// set box size to 0
+			newBoxSize := fyne.Size{Width: minBarWidth, Height: progressSize.Height}
+			p.bar.Resize(newBoxSize)
+		} else if barPos.X+barWidth > progressSize.Width {
+			// crop to the end of the bar
+			barWidth = progressSize.Width - barPos.X
+			newBoxSize := fyne.Size{Width: barWidth, Height: progressSize.Height}
+			p.bar.Resize(newBoxSize)
+		}
+	}
+
+	p.bar.Move(barPos)
+}
+
+// Layout the components of the progress bar widget
+func (p *infProgressRenderer) Layout(size fyne.Size) {
+	// set height of progress bar
+	p.updateBar()
+}
+
+// ApplyTheme is called when the progress bar may need to update it's look
+func (p *infProgressRenderer) ApplyTheme() {
+	p.bar.FillColor = theme.PrimaryColor()
+
+	p.Refresh()
+}
+
+func (p *infProgressRenderer) BackgroundColor() color.Color {
+	return theme.ButtonColor()
+}
+
+func (p *infProgressRenderer) Refresh() {
+	p.updateBar()
+	canvas.Refresh(p.progress)
+}
+
+func (p *infProgressRenderer) Objects() []fyne.CanvasObject {
+	return p.objects
+}
+
+// InfProgressBar widget creates a horizontal panel that indicates waiting indefinitely
+// An infinite progress bar loops 0% -> 100% until Stop() is called
+type InfProgressBar struct {
+	baseWidget
+
+	ticker *time.Ticker
+}
+
+// Resize sets a new size for a widget.
+// Note this should not be used if the widget is being managed by a Layout within a Container.
+func (p *InfProgressBar) Resize(size fyne.Size) {
+	p.resize(size, p)
+}
+
+// Move the widget to a new position, relative to it's parent.
+// Note this should not be used if the widget is being managed by a Layout within a Container.
+func (p *InfProgressBar) Move(pos fyne.Position) {
+	p.move(pos, p)
+}
+
+// MinSize returns the smallest size this widget can shrink to
+func (p *InfProgressBar) MinSize() fyne.Size {
+	return p.minSize(p)
+}
+
+// Show this widget, if it was previously hidden
+func (p *InfProgressBar) Show() {
+	p.Start()
+	p.show(p)
+}
+
+// Hide this widget, if it was previously visible
+func (p *InfProgressBar) Hide() {
+	p.Stop()
+	p.hide(p)
+}
+
+// Start the infinite progress bar background thread to update it continuously
+func (p *InfProgressBar) Start() {
+	if p.ticker == nil {
+		go p.infiniteProgressLoop()
+	}
+}
+
+// Stop the infinite progress goroutine and sets value to the Max
+func (p *InfProgressBar) Stop() {
+	if p.ticker != nil {
+		p.ticker.Stop()
+		p.ticker = nil
+	}
+}
+
+// internal loop called with `go infiniteProgressLoop()`
+// updates the infinite-style progress bar
+// can be exited by calling ProgressBar.StopInfiniteProgress()
+func (p *InfProgressBar) infiniteProgressLoop() {
+	defer p.Stop()
```

Comment:
I don't think this line is needed - as Stop() must already have been called to exit the ticker range?

Reply:
`Stop()` checks to make sure the ticker is running (`p.ticker != nil`) before doing anything.  My thinking was that calling it multiple times is a lot better than not calling it at all because of the low overhead.

Reply:
Not sure... I'd go for calling it exactly the right number of times ;)

Reply:
Looking at it with a fresh glass of beer, and you're right in that the ticker needs to stop before the defer would get hit, rendering the call useless always.  I'll remove this.